checkpoint/runner_test.go

```diff
@@ -0,0 +1,189 @@
+	"encoding/json"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/spf13/afero"
+	"github.com/stretchr/testify/require"
+
+	"github.com/spacemeshos/go-spacemesh/checkpoint"
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/log/logtest"
+	"github.com/spacemeshos/go-spacemesh/sql"
+	"github.com/spacemeshos/go-spacemesh/sql/accounts"
+	"github.com/spacemeshos/go-spacemesh/sql/atxs"
+)
+
+func TestMain(m *testing.M) {
+	types.SetLayersPerEpoch(3)
+	res := m.Run()
+	os.Exit(res)
+}
+
+type createAtxOpt func(*types.ActivationTx)
+
+func withSmesherID(nid types.NodeID) createAtxOpt {
+	return func(atx *types.ActivationTx) {
+		atx.SmesherID = nid
+	}
+}
+
+func withPublishEpoch(epoch types.EpochID) createAtxOpt {
+	return func(atx *types.ActivationTx) {
+		atx.PublishEpoch = epoch
+	}
+}
+
+func withSequence(seq uint64) createAtxOpt {
+	return func(atx *types.ActivationTx) {
+		atx.Sequence = seq
+		atx.PublishEpoch = types.EpochID(uint32(seq + 1))
+	}
+}
+
+func newAtx(opts ...createAtxOpt) (*types.VerifiedActivationTx, error) {
+	atx := &types.ActivationTx{
+		InnerActivationTx: types.InnerActivationTx{
+			NIPostChallenge: types.NIPostChallenge{
+				PrevATXID: types.RandomATXID(),
+			},
+			Coinbase: types.GenerateAddress(types.RandomBytes(20)),
+			NumUnits: 2,
+		},
+	}
+	for _, opt := range opts {
+		opt(atx)
+	}
+	if atx.Sequence == 0 {
+		atx.CommitmentATX = &types.ATXID{1, 2, 3}
+		nonce := types.VRFPostIndex(11)
+		atx.VRFNonce = &nonce
+	}
+	atx.SetEffectiveNumUnits(atx.NumUnits)
+	atx.SetReceived(time.Now().Local())
+	return atx.Verify(0, 1)
+}
+
+func genAtxSeq(t *testing.T, nid types.NodeID, n int) []*types.VerifiedActivationTx {
+	var seq []*types.VerifiedActivationTx
+	for i := 0; i < n; i++ {
+		vatx, err := newAtx(withSmesherID(nid), withSequence(uint64(i)), withPublishEpoch(types.EpochID(i+1)))
+		require.NoError(t, err)
+		seq = append(seq, vatx)
+	}
+	return seq
+}
+
+func genAccountSeq(address types.Address, n int) []*types.Account {
+	var seq []*types.Account
+	for i := 1; i <= n; i++ {
+		seq = append(seq, &types.Account{
+			Address:         address,
+			Layer:           types.LayerID(uint32(i)),
+			Balance:         uint64(i),
+			NextNonce:       uint64(i + 1),
+			TemplateAddress: &types.Address{1},
+			State:           []byte{1, 3},
+		})
+	}
+	return seq
+}
+
+func toShortAtx(v *types.VerifiedActivationTx, cmt *types.ATXID, nonce *types.VRFPostIndex) checkpoint.ShortAtx {
+	return checkpoint.ShortAtx{
+		ID:             hex.EncodeToString(v.ID().Bytes()),
+		Epoch:          v.PublishEpoch.Uint32(),
+		CommitmentAtx:  hex.EncodeToString(cmt.Bytes()),
+		VrfNonce:       uint64(*nonce),
+		NumUnits:       v.NumUnits,
+		BaseTickHeight: v.BaseTickHeight(),
+		TickCount:      v.TickCount(),
+		PublicKey:      hex.EncodeToString(v.SmesherID.Bytes()),
+		Sequence:       v.Sequence,
+		Coinbase:       hex.EncodeToString(v.Coinbase.Bytes()),
+	}
+}
+
+func TestRunner_Generate(t *testing.T) {
```

Comment:
i would try to define some table based framefork so that it can be easily extended. in this case inputs are atxs and accounts, and output is encoded json, right? so maybe it makes sense to have a table like:


```
{
   "empty"
    nil,
    nil,
    "accounts: [], atxs: []",
}
{
    "only atxs",
    []*types.VerifiedAtx{}
}
```

Reply:
tried my best to re-write the test setup. now it listed the atxs and accounts explicitly.
linter now doesn't allow unnamed struct fields.